File: LeetCode/123. Best Time to Buy and Sell Stock III.cpp
/***  KNOWLEDGE IS POWER  ***/

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        int minn = prices[0],  maxx = prices[n-1], prev = 0, ans = 0;
        vector<int> pref(n, 0);
        
        for(int i = 1; i < n; i++) {
            pref[i] = max(pref[i-1], prices[i] - minn);
            minn = min(minn, prices[i]);
        }

        for(int i = n-2; i >= 0; i--) {
            prev = max(prev, maxx - prices[i]);
            ans = max(pref[i] + prev, ans);
            maxx = max(maxx, prices[i]);   
        }
        return ans;
    }
};

File: LeetCode/123. Best Time to Buy and Sell Stock III.py
### KNOWLEDGE IS POWER  ###

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        s0 = 0
        h1 = s1 = h2 = s2 = float('-inf')
        for p in prices:
            h1, s1, h2, s2 = max(h1, s0 - p), max(s1, h1 + p), max(h2, s1 - p), max(s2, h2 + p)
        return max(s0, s1, s2)
